rp2: Fix USB PLL glitch during wake from light sleep.

projectgus opened the pull request:
Follow-up to #15111, that change works most of the time but has an intermittent bug where USB doesn't resume as expected after waking from light sleep. Thanks to @GitHubsSilverBullet who reported it here: https://github.com/orgs/micropython/discussions/14401#discussioncomment-9675017

Turns out waking calls `clocks_init()` which will re-initialise the USB PLL. Most of the time this is OK but occasionally it seems like the clock glitches the USB peripheral and it stops working until the next hard reset.

Adds a `machine.lightsleep()` test that consistently hangs in the first two dozen iterations of a single test run on rp2 without this fix. Passed over 100 full test runs in a row with this fix.

The test is currently rp2-only as it seems similar lightsleep USB issues exist on other ports (both pyboard and ESP32-S3 native USB don't send any data to the host after waking, until they receive something from the host first.)

EDIT: Added one more small commit to always disable USB if there is no timeout (i.e. entering DORMANT mode where the external oscillator is disabled.)

*This work was funded through GitHub Sponsors.*

dpgeorge commented:
Maybe we should report this to pico-sdk, and potentially they can add this argument to `clocks_init()`?

But for now, patching it here is OK.

Can you please split out the addition of our copy of `clocks_init()` to a separate commit and update the top-level LICENSE?  See 8438c8790c0ad11545371a07f0b0ce4c65c62101

projectgus committed:
rp2/clocks_extra: Implement custom clocks_init function.

Adapts pico-sdk clocks_init() into clocks_init_optional_usb() which takes
an argument to initialise USB clocks or not.

To avoid a code size increase the SDK clocks_init() function is linker
wrapped to become clocks_init_optional_usb(true).

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
rp2: Fix USB PLL glitch during wake from light sleep.

Follow-up to a84c7a0ed93, this commit works most of the time but has an
intermittent bug where USB doesn't resume as expected after waking from
light sleep.

Turns out waking calls clocks_init() which will re-initialise the USB PLL.
Most of the time this is OK but occasionally it seems like the clock
glitches the USB peripheral and it stops working until the next hard reset.

Adds a machine.lightsleep() test that consistently hangs in the first
two dozen iterations on rp2 without this fix. Passed over 100 times in a
row with this fix.

The test is currently rp2-only as it seems similar lightsleep USB issues
exist on other ports (both pyboard and ESP32-S3 native USB don't send any
data to the host after waking, until they receive something from the host
first.)

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
> Can you please split out the addition of our copy of `clocks_init()` to a separate commit and update the top-level LICENSE? See [8438c87](https://github.com/micropython/micropython/commit/8438c8790c0ad11545371a07f0b0ce4c65c62101)

No worries, done. PTAL!

projectgus committed:
rp2: Don't disable USB if going to DORMANT mode.

In this mode, XOSC is stopped so can't really keep
the USB PLL enabled.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
Added one more small commit to always disable USB if going to DORMANT mode with no external oscillator.

Have also raised a feature request with pico-sdk, see link above.

dpgeorge commented:
All looks good, thank you!
